### PR TITLE
add hbase rest

### PR DESCRIPTION
**Pull request by jetoile**
This PR propose a way to have a hbase rest.

It should be useful for knox integration.

However I had no other choice that duplicate the class RESTServer from HBase because of method visibility.

For the start method, I just copied the main method from RESTServer

Do you think there is a better way?

Regards,

### CODE REVIEW

File: hadoop-mini-clusters-hbase/src/main/java/com/github/sakserv/minicluster/impl/HbaseRestLocalCluster.java [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+            if (hbaseRestLocalCluster.hbaseRestPort == null) {
+                throw new IllegalArgumentException("ERROR: Missing required config: HBase Rest Port");
+            }
+        }
+    }
+
+    @Override
+    public void start() throws Exception {
+        VersionInfo.logVersion();
+        Configuration conf = HBaseConfiguration.create();
```
Comment (Treydone) [inline hadoop-mini-clusters-hbase/src/main/java/com/github/sakserv/minicluster/impl/HbaseRestLocalCluster.java]: Can be dangerous I think, I will load all the *-site.xml and *-core.xml by default from the classpath. Maybe a boolean "default load from classpath"?

File: hadoop-mini-clusters-hbase/src/main/java/com/github/sakserv/minicluster/impl/HbaseRestLocalCluster.java [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+            context.addFilter(authFilter, "/*", 1);
+        }
+
+        // Load filters from configuration.
+//        String[] filterClasses = servlet.getConfiguration().getStrings(RESTServer.FILTER_CLASSES, ArrayUtils.EMPTY_STRING_ARRAY);
+//        for (String filter : filterClasses) {
+//            filter = filter.trim();
+//            context.addFilter(Class.forName(filter), "/*", 0);
+//        }
+        RESTServer.addCSRFFilter(context, conf);
```
Comment (Treydone) [inline hadoop-mini-clusters-hbase/src/main/java/com/github/sakserv/minicluster/impl/HbaseRestLocalCluster.java]: Is it realy needed for a dev tools?

File: hadoop-mini-clusters-hbase/src/main/java/org/apache/hadoop/hbase/rest/RESTServer.java [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+ * Main class for launching REST gateway as a servlet hosted by Jetty.
+ * <p>
+ * The following options are supported:
+ * <ul>
+ * <li>-p --port : service port</li>
+ * <li>-ro --readonly : server mode</li>
+ * </ul>
+ */
+@InterfaceAudience.LimitedPrivate(HBaseInterfaceAudience.TOOLS)
+public class RESTServer implements Constants {
```
Comment (Treydone) [inline hadoop-mini-clusters-hbase/src/main/java/org/apache/hadoop/hbase/rest/RESTServer.java]: For method visibility issues, a wrapper class (RESTServerWrapper) in the package org.apache.hadoop.hbase.rest, extending RESTServer and overriding the protected/package protected methods by calling super.xxx should be more clear, what do you think?

Comment (jetoile) [inline hadoop-mini-clusters-hbase/src/main/java/org/apache/hadoop/hbase/rest/RESTServer.java]: indeed, you are right ;)

File: hadoop-mini-clusters-hbase/src/main/java/com/github/sakserv/minicluster/impl/HbaseRestLocalCluster.java (hunk 2 above)
Comment (jetoile) [inline hadoop-mini-clusters-hbase/src/main/java/com/github/sakserv/minicluster/impl/HbaseRestLocalCluster.java]: Indeed. If security is disabled, no more problem with RESTServer... ;)
@sakserv : What do you think ?

Comment (sakserv) [inline hadoop-mini-clusters-hbase/src/main/java/com/github/sakserv/minicluster/impl/HbaseRestLocalCluster.java]: @jetoile - Disabling security works for me. We can reintroduce it in the future if needed.

**Comment by sakserv** [comment]
Thanks @jetoile for the awesome contribution! and thank you @Treydone for the review. @jetoile it seems there is one remaining comment to address from @Treydone. If you could address that, I'll get this merged today and will work to get a new release out this week.

**Comment by jetoile** [comment]
Thx @sakserv for your reply.
I have an other question too. With @Treydone (who is a friend ;) ), we thought that it could be nice to integrate the rest gateway into hbase.
However, as some options are mandatory only for rest hbase, it could be bad.
But we could do that with something like:

```
hbaseLocalCluster = new HbaseLocalCluster.Builder()
                ...
                .activeRestGateway()
                    .setHbaseRestHost(propertyParser.getProperty(ConfigVars.HBASE_REST_HOST_KEY))
                    .setHbaseRestReadOnly(
                            Boolean.valueOf(propertyParser.getProperty(ConfigVars.HBASE_REST_READONLY_KEY)))
                    .setHbaseRestThreadMax(
                            Integer.valueOf(propertyParser.getProperty(ConfigVars.HBASE_REST_THREADMAX_KEY)))
                    .setHbaseRestThreadMin(
                            Integer.valueOf(propertyParser.getProperty(ConfigVars.HBASE_REST_THREADMIN_KEY)))
                    .build()
                .build();
```

What do you think?

**Comment by sakserv** [comment]
@jetoile - seems like a good idea to me. I don't think we necessarily need a new Builder, but I'm not opposed to that pattern either. ```HbaseLocalCluster#validateObject``` is responsible for checking *required* args. If you leave these args out of that check, they will be treated as optional.

**Comment by jetoile** [comment]
I just push an update with a other builder.
This PR resolve the comment @Treydone has done to me too.
Tell me what you think about.
If you want I can squash all the commits.

Regards,

Khanh

**Comment by sakserv** [comment]
LGTM - once the CI run completes, I'll go ahead and merge this. Thanks again @jetoile and @Treydone!!

**Comment by sakserv** [comment]
These changes have been included in 0.1.10 which was just released. Thanks again!

**Pull request merged**